reservation_units/utils/reservation_unit_state_helper.py

```diff
@@ -0,0 +1,124 @@
+def is_scheduled_publishing(reservation_unit: ReservationUnit) -> bool:
+    now = datetime.datetime.now(tz=get_default_timezone())
+    if is_draft(reservation_unit) or is_archived(reservation_unit):
+        return False
+
+    return (
+        reservation_unit.publish_begins and now < reservation_unit.publish_begins
+    ) or (reservation_unit.publish_ends and now >= reservation_unit.publish_ends)
+
+
+def get_is_scheduled_publishing_query() -> Q:
+    now = datetime.datetime.now(tz=get_default_timezone())
+    return Q(is_archived=False, is_draft=False) & (
+        Q(publish_begins__isnull=False, publish_begins__gt=now)
+        | Q(publish_ends__isnull=False, publish_ends__lte=now)
+    )
+
+
+# SCHEDULED_RESERVATION
+def is_scheduled_reservation(reservation_unit: ReservationUnit) -> bool:
+    now = datetime.datetime.now(tz=get_default_timezone())
+    if is_draft(reservation_unit) or is_archived(reservation_unit):
+        return False
+    return (
+        reservation_unit.reservation_begins
+        and now < reservation_unit.reservation_begins
+    ) or (
+        reservation_unit.reservation_ends and now >= reservation_unit.reservation_ends
+    )
+
+
+def get_is_scheduled_reservation_query() -> Q:
+    now = datetime.datetime.now(tz=get_default_timezone())
+    return Q(reservation_begins__isnull=False, reservation_begins__gt=now) | Q(
+        reservation_ends__isnull=False, reservation_ends__lte=now
+    )
+
+
+# PUBLISHED
+def is_published(reservation_unit: ReservationUnit) -> bool:
+    return (
+        not is_draft(reservation_unit)
+        and not is_archived(reservation_unit)
+        and not is_scheduled_publishing(reservation_unit)
+        and not is_scheduled_reservation(reservation_unit)
+    )
+
+
+def get_is_published_query() -> Q:
+    now = datetime.datetime.now(tz=get_default_timezone())
+    return (
+        Q(is_archived=False)
+        & Q(is_draft=False)
+        & (Q(publish_begins__isnull=True) | Q(publish_begins__lte=now))
+        & (Q(publish_ends__isnull=True) | Q(publish_ends__gt=now))
+        & (Q(reservation_begins__isnull=True) | Q(reservation_begins__lte=now))
+        & (Q(reservation_ends__isnull=True) | Q(reservation_ends__gt=now))
+    )
+
+
+def get_state(reservation_unit: ReservationUnit) -> ReservationUnitState:
+    """Figure out the state of the Reservation Unit"""
+    if is_archived(reservation_unit):
+        return ReservationUnitState.ARCHIVED
+    if is_draft(reservation_unit):
+        return ReservationUnitState.DRAFT
+    if is_scheduled_publishing(reservation_unit):
+        return ReservationUnitState.SCHEDULED_PUBLISHING
+    if is_scheduled_reservation(reservation_unit):
+        return ReservationUnitState.SCHEDULED_RESERVATION
+    if is_published(reservation_unit):
+        return ReservationUnitState.PUBLISHED
+
+    raise ValueError(
+        f"Unable to detect the state of the reservation unit {reservation_unit.id}"
+    )
+
+
+def get_state_query(state: str) -> Q:
```

Comment:
This whole logic could be wrapped into a class instead of separate functions :shrug: At least there's a some functions smelling like private methods for a class :thinking:

Reply:
Changed to class and made most of the functions private methods. Feel free to check if my approach to "classification" follows the best practices 🥸